internal/gosourcefile/gosourcefile.go

```diff
@@ -1,40 +1,58 @@
 package gosourcefile
 
 import (
-	"fmt"
 	"go/ast"
-	"go/parser"
-	"go/token"
+	"go/types"
+	"path"
 
 	"github.com/gtramontina/ooze/internal/goinfectedfile"
 	"github.com/gtramontina/ooze/viruses"
+	"golang.org/x/tools/go/packages"
 )
 
 type GoSourceFile struct {
+	root         string
 	relativePath string
 	rawContent   []byte
+	pkg          *packages.Package
+	typeInfo     *types.Info
 }
 
-func New(relativePath string, rawContent []byte) *GoSourceFile {
-	return &GoSourceFile{
+func New(root string, relativePath string, rawContent []byte) *GoSourceFile {
+	return &GoSourceFile{ //nolint:exhaustruct // rest will be filled later
+		root:         root,
 		relativePath: relativePath,
 		rawContent:   rawContent,
 	}
 }
 
 func (f *GoSourceFile) Incubate(virus viruses.Virus) []*goinfectedfile.GoInfectedFile {
-	fileSet := token.NewFileSet()
+	f.loadTypeInfo()
 
-	fileTree, err := parser.ParseFile(fileSet, f.relativePath, f.rawContent, parser.ParseComments|parser.AllErrors)
-	if err != nil {
-		panic(fmt.Errorf("failed parsing file '%s': %w", f.relativePath, err))
+	fullPath := path.Join(f.root, f.relativePath)
+
+	idx := -1
+
+	for i, path := range f.pkg.CompiledGoFiles {
+		if path == fullPath {
+			idx = i
+
+			break
+		}
 	}
 
+	if idx < 0 {
+		return nil
+	}
```

Comment:
Will this ever happen? I'm thinking that if it ever happens, it means that something has actually gone wrong pretty badly, no? I'd consider blowing up with a panic here. WDYT?

Reply:
Yeah sounds good to me.